src/main/java/net/sf/mzmine/modules/peaklistmethods/filtering/groupms2/GroupMS2Task.java

```diff
@@ -0,0 +1,164 @@
+import net.sf.mzmine.parameters.parametertypes.tolerances.MZTolerance;
+import net.sf.mzmine.parameters.parametertypes.tolerances.RTTolerance;
+import net.sf.mzmine.taskcontrol.AbstractTask;
+import net.sf.mzmine.taskcontrol.TaskStatus;
+
+/**
+ * Filters out peak list rows.
+ */
+public class GroupMS2Task extends AbstractTask {
+
+  // Logger.
+  private static final Logger LOG = Logger.getLogger(GroupMS2Task.class.getName());
+  // Peak lists.
+  private final MZmineProject project;
+  // Processed rows counter
+  private int processedRows, totalRows;
+  // Parameters.
+  private final ParameterSet parameters;
+  private PeakList list;
+  private RTTolerance rtTol;
+  private MZTolerance mzTol;
+  private boolean limitRTByFeature;
+
+  /**
+   * Create the task.
+   *
+   * @param list peak list to process.
+   * @param parameterSet task parameters.
+   */
+  public GroupMS2Task(final MZmineProject project, final PeakList list,
+      final ParameterSet parameterSet) {
+
+    // Initialize.
+    this.project = project;
+    parameters = parameterSet;
+    rtTol = parameters.getParameter(GroupMS2Parameters.rtTol).getValue();
+    mzTol = parameters.getParameter(GroupMS2Parameters.mzTol).getValue();
+    limitRTByFeature = parameters.getParameter(GroupMS2Parameters.limitRTByFeature).getValue();
+    this.list = list;
+    processedRows = 0;
+    totalRows = 0;
+  }
+
+  @Override
+  public double getFinishedPercentage() {
+
+    return totalRows == 0 ? 0.0 : (double) processedRows / (double) totalRows;
+  }
+
+  @Override
+  public String getTaskDescription() {
+
+    return "Adding all MS2 scans to their features in list " + list.getName();
+  }
+
+  @Override
+  public void run() {
+
+    try {
+      setStatus(TaskStatus.PROCESSING);
+
+      totalRows = list.getNumberOfRows();
+      // for all features
+      for (PeakListRow row : list.getRows()) {
+        for (Feature f : row.getPeaks()) {
+          if (getStatus() == TaskStatus.ERROR)
+            return;
+          if (isCanceled())
+            return;
+
+          RawDataFile raw = f.getDataFile();
+          IntArrayList scans = new IntArrayList();
+          int best = f.getRepresentativeScanNumber();
+          double frt = f.getRT();
+          double fmz = f.getMZ();
+          Range<Double> rtRange = f.getRawDataPointsRTRange();
+          int i = best;
+          // left
+          while (i > 1) {
+            try {
+              i--;
+              Scan scan = raw.getScan(i);
+              if ((!limitRTByFeature || rtRange.contains(scan.getRetentionTime()))
+                  && rtTol.checkWithinTolerance(frt, scan.getRetentionTime())) {
+                if (scan.getPrecursorMZ() != 0
+                    && mzTol.checkWithinTolerance(fmz, scan.getPrecursorMZ()))
+                  scans.add(i);
+              } else {
+                // end of loop - out of tolerance
+                break;
+              }
+            } catch (Exception e) {

```

Comment:
At least add e.printStackTrace(), you should never swallow exceptions without warning.

Reply:
Yeah changed it to a null checking. 
The RawDataFile.getScan(i) method should be @Nullable 
The user can enter any index which is out of range or just not in the HashTable (the scan numbers do not have to be contiguous).